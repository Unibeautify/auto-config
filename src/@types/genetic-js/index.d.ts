// declare module "genetic-js";

declare namespace GeneticJs {
  export interface Configuration {
    size: number;
    crossover: number;
    mutation: number;
    iterations: number;
    fittestAlwaysSurvives: boolean;
    maxResults: number;
    webWorkers: boolean;
    skip: number;
  }
  export type Population<Entity> = Array<{
    fitness: number;
    entity: Entity;
  }>;
  export interface Stats {
    maximum: number;
    minimum: number;
    mean: number;
    stdev: number;
  }
  export interface Genetic<Entity> {
    optimize: OptimizeFun;
    seed: () => Entity;
    mutate: (entity: Entity) => Entity;
    crossover: (mother: Entity, father: Entity) => [Entity, Entity];
    fitness: (entity: Entity) => number;
    generation: (
      pop: Population<Entity>,
      generation: number,
      stats: Stats
    ) => boolean;
    notification: (
      pop: Population<Entity>,
      generation: number,
      stats: Stats,
      isFinished: boolean
    ) => void;
    select1: SingleSelection<Entity>;
    select2: PairWiseSelection<Entity>;
    evolve(config: Partial<Configuration>, userData: any): void;
  }

  export type SingleSelection<Entity> = (pop: Population<Entity>) => Entity;
  export type PairWiseSelection<Entity> = (
    pop: Population<Entity>
  ) => [Entity, Entity];
  export interface Select1Map<Entity> {
    Tournament2: GeneticJs.SingleSelection<Entity>;
    Tournament3: GeneticJs.SingleSelection<Entity>;
    Fittest: GeneticJs.SingleSelection<Entity>;
    Random: GeneticJs.SingleSelection<Entity>;
    RandomLinearRank: GeneticJs.SingleSelection<Entity>;
    Sequential: GeneticJs.SingleSelection<Entity>;
  }
  export interface Select2Map<Entity> {
    Tournament2: GeneticJs.PairWiseSelection<Entity>;
    Tournament3: GeneticJs.PairWiseSelection<Entity>;
    Random: GeneticJs.PairWiseSelection<Entity>;
    RandomLinearRank: GeneticJs.PairWiseSelection<Entity>;
    Sequential: GeneticJs.PairWiseSelection<Entity>;
    FittestRandom: GeneticJs.PairWiseSelection<Entity>;
  }

  export type OptimizeFun = (fitnessA: number, fitnessB: number) => boolean;
  export interface Optimize {
    Minimize: OptimizeFun;
    Maximize: OptimizeFun;
  }
}

declare module "genetic-js" {
  export function create<Entity>(): GeneticJs.Genetic<Entity>;
  export const Optimize: GeneticJs.Optimize;
  export const Select1: GeneticJs.Select1Map<any>;
  export const Select2: GeneticJs.Select2Map<any>;
  namespace cosmiconfig {

  }
}
