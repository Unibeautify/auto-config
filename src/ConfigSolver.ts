// import * as Genetic from "@glavin001/genetic-js";
import * as fastLevenshtein from "fast-levenshtein";
import * as _ from "lodash";
// import * as stringify from "json-stable-stringify";

import { UnibeautifyGenetic, Options, Entity } from "./UnibeautifySolver";

export class ConfigSolver extends UnibeautifyGenetic {
  private readonly optionsCount: number;
  private readonly originalTextVariations: Promise<string>[];

  constructor(options: Options) {
    const { userData } = options;
    if (userData.originalText !== userData.desiredText) {
      throw new Error(
        "This solver should only be used when original and desired text are the same."
      );
    }
    super(options);
    this.optionsCount = this.optionKeys.length;
    console.log(JSON.stringify(this.optionKeys, null, 2));
    const variationsCount = 30;
    this.originalTextVariations = [];
    for (let currSeed = 0; currSeed < variationsCount; ++currSeed) {
      this.originalTextVariations.push(
        this.beautify(this.originalText, this.seed())
      );
    }
  }

  protected internalFitness(entity: Entity): Promise<number> {
    return Promise.all(this.originalTextVariations)
      .then(originalTextVariations => {
        return Promise.all(
          originalTextVariations.map(text =>
            this.internalIndividualFitness(text, entity)
          )
        );
      })
      .then(
        scores => scores.reduce((sum, curr) => sum + curr, 0) / scores.length
      );
  }

  protected internalIndividualFitness(
    originalText: string,
    entity: Entity
  ): Promise<number> {
    // Minimize diff count
    // Maximum # of options used ==> more options used then smaller score (smaller is better)
    return this.beautify(originalText, entity).then(
      (beautifiedText: string) => {
        // const diffCount = fastLevenshtein.get(this.desiredText, beautifiedText);
        const diffCount = this.diffCount(beautifiedText, this.desiredText);
        // const diffFitness: number = diffCount * 10000;
        // const numOfBeautifiers = (<any>entity.options).beautifiers.length;
        // if (diffCount !== 0) {
        //   return diffFitness;
        // }
        // const beautifierFitness: number =
        //   Math.max(0, numOfBeautifiers - 1) * 1000;

        // const ignoreOptionKeys = [
        //   "beautifiers",
        //   //   "indent_size",
        //   //   "indent_style",
        // ];
        // const optionsCount: number = Math.max(
        //   0,
        //   _.without(_.keys(entity.options), ...ignoreOptionKeys).length - 1
        // );
        const optionsCount = this.optionsUsed(entity);

        // return beautifierFitness + optionsFitness;
        // const optionsFitness = optionsCount >= this.optionsCount ? 0 : 1 / (this.optionsCount - optionsCount);
        // const optionsFitness = 1 / (optionsCount + 1);
        // const optionsFitness = 1 / (this.optionsCount / (optionsCount + 1) + 1) * 100;
        // if (optionsCount <= 0) {
        //   // return Infinity;
        //   return 10000000;
        // }

        const diffPerc: number =
          diffCount / Math.max(this.desiredText.length, beautifiedText.length);
        const diffFitness: number = diffPerc * 10000;

        // const optionsPerc: number = optionsCount / this.optionsCount;
        // const optionsFitness: number = 1 / optionsPerc * 100;

        // const optionsFitness = Math.max(0, this.optionsCount - optionsCount);

        const optionsPerc: number =
          Math.max(0, this.optionsCount - optionsCount) / this.optionsCount;
        const optionsFitness: number = optionsPerc * 100;

        return diffFitness + optionsFitness;
      }
    );
  }

  protected diffCount(src: string, dest: string): number {
    return fastLevenshtein.get(src, dest);
  }

  protected optionsUsed(entity: Entity): number {
    const ignoreOptionKeys = [
      "beautifiers",
      //   "indent_size",
      //   "indent_style",
    ];
    const optionsCount: number = Math.max(
      0,
      _.without(_.keys(entity.options), ...ignoreOptionKeys).length - 1
    );
    return optionsCount;
  }

  // public shouldContinue({ population }: Genetic.GeneticState<Entity>) {
  //   const bestFitness = population[0].fitness;
  //   const isPerfect = bestFitness === 0;
  //   if (isPerfect) {
  //     return false;
  //   }
  //   // const textIsTheSame = Math.floor(bestFitness / 10000) === 0;
  //   // if (textIsTheSame) {
  //   //   const bestEntities = population.filter(
  //   //     entity => entity.fitness === bestFitness
  //   //   );
  //   //   const threshold = this.configuration.size * 0.5;
  //   //   if (bestEntities.length > threshold) {
  //   //     const entityJson = stringify(population[0].entity);
  //   //     const bestAreSame = bestEntities.every(
  //   //       ({ entity }) => stringify(entity) === entityJson
  //   //     );
  //   //     if (bestAreSame) {
  //   //       return false;
  //   //     }
  //   //   }
  //   // }
  //   return true;
  // }
}
