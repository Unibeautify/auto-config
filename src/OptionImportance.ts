import {
  BeautifierOptionName,
  OptionsRegistry,
  newUnibeautify,
  Unibeautify,
  Beautifier,
  Language,
  Option,
  Options,
  OptionValues,
} from "unibeautify";
import { UnibeautifyGenetic, Entity } from "./UnibeautifySolver";
import * as _ from "lodash";

export class ImportantOptionsRegistryBuilder {
  private readonly supportedOptionsRegistry: OptionsRegistry;
  protected readonly optionKeys: BeautifierOptionName[];

  constructor(private solver: UnibeautifyGenetic) {
    this.supportedOptionsRegistry = this.solver.optionsRegistry;
    this.optionKeys = Object.keys(
      this.supportedOptionsRegistry
    ).sort() as any[];
  }
  //   private readonly unibeautify: Unibeautify;
  //   private readonly language: Language;

  //   constructor(private userData: UserData) {
  //     this.unibeautify = newUnibeautify();
  //     this.initializeUnibeautify();
  //     this.language = this.getLanguage();
  //   }

  //   private initializeUnibeautify() {
  //     this.unibeautify.loadBeautifiers(this.beautifiers);
  //   }
  //   private get beautifiers(): Beautifier[] {
  //     return this.userData.beautifiers;
  //   }

  //   private getLanguage(): Language {
  //     return this.unibeautify.findLanguages({
  //       name: this.languageName,
  //     })[0];
  //   }

  //   private get languageName(): string {
  //     return this.userData.language;
  //   }

  //   protected get optionKeys(): BeautifierOptionName[] {
  //     return Object.keys(this.supportedOptionsRegistry).sort() as any[];
  //   }

  //   private get optionsRegistry(): OptionsRegistry {
  //     return this.unibeautify.getOptionsSupportedForLanguage(this.language);
  //   }

  public buildImportantOptionsRegistry(
    optionValues: OptionValues = {}
  ): Promise<OptionsRegistry> {
    return Promise.all(
      this.optionKeys.map((optionKey: BeautifierOptionName) => {
        return this.optionIsImportant(optionKey, optionValues);
      })
    )
      .then(shouldIncludeOptions => {
        return this.optionKeys.filter(
          (optionKeys, index) => shouldIncludeOptions[index]
        );
      })
      .then(includedOptions => {
        return _.pick(this.supportedOptionsRegistry, includedOptions);
      });
  }

  private optionIsImportant(
    optionKey: BeautifierOptionName,
    optionValues: OptionValues
  ): Promise<boolean> {
    const option = this.supportedOptionsRegistry[optionKey];
    const values = this.valuesForOption(option);
    return Promise.all(
      values.map(value =>
        this.fitness(optionKey, value, optionValues).then(fitness => {
          console.log("fitness", optionKey, fitness);
          return fitness;
        })
      )
    ).then(fitnesses => {
      const firstVal = fitnesses[0];
      const shouldInclude = !fitnesses.every(f => f === firstVal);
      console.log(
        "fitnesses",
        optionKey,
        shouldInclude,
        _.zipObject(values, fitnesses)
      );
      return shouldInclude;
    });
  }

  private valuesForOption(option: Option): any[] {
    return this.solver.exampleValues(option);
  }

  private fitness(
    optionKey: BeautifierOptionName,
    value: any,
    optionValues: OptionValues
  ): Promise<number> {
    // FIXME: Beautifier may support option and not language
    const beautifiers = this.solver.unibeautify.getBeautifiersSupportingOption(
      optionKey
    );
    const beautifierNames: string[] = beautifiers.map(
      beautifier => beautifier.name
    );
    const entity: Entity = {
      options: {
        ...({
          beautifiers: beautifierNames,
        } as any),
        indent_size: 2,
        indent_style: "space",
        ...optionValues,
        [optionKey]: value,
      },
    };
    return this.solver.fitness(entity);
  }
}

// export interface UserData {
//   beautifiers: Beautifier[];
//   language: string;
//   desiredText: string;
// }
