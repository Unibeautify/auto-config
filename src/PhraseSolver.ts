import * as Genetic from "@glavin001/genetic-js";

type Entity = string;
interface UserData {
  solution: Entity;
}
// const genetic = Genetic.create<Entity, UserData>();
class PhraseGenetic extends Genetic.Genetic<Entity, UserData> {
  optimize = Genetic.Optimize.Maximize;
  select1 = Genetic.Select1.Tournament2;
  select2 = Genetic.Select2.Tournament2;

  seed() {
    function randomString(len: number) {
      let text = "";
      const charset = "abcdefghijklmnopqrstuvwxyz0123456789";
      for (var i = 0; i < len; i++)
        text += charset.charAt(Math.floor(Math.random() * charset.length));

      return text;
    }

    // create random strings that are equal in length to solution
    return randomString(this.userData["solution"].length);
  }

  mutate(entity: Entity) {
    function replaceAt(str: string, index: number, character: string) {
      return (
        str.substr(0, index) + character + str.substr(index + character.length)
      );
    }

    // chromosomal drift
    var i = Math.floor(Math.random() * entity.length);
    return replaceAt(
      entity,
      i,
      String.fromCharCode(
        entity.charCodeAt(i) + (Math.floor(Math.random() * 2) ? 1 : -1)
      )
    );
  }

  crossover(mother: Entity, father: Entity): [Entity, Entity] {
    // two-point crossover
    var len = mother.length;
    var ca = Math.floor(Math.random() * len);
    var cb = Math.floor(Math.random() * len);
    if (ca > cb) {
      var tmp = cb;
      cb = ca;
      ca = tmp;
    }

    var son =
      father.substr(0, ca) + mother.substr(ca, cb - ca) + father.substr(cb);
    var daughter =
      mother.substr(0, ca) + father.substr(ca, cb - ca) + mother.substr(cb);

    return [son, daughter];
  }

  fitness(entity: Entity) {
    var fitness = 0;

    var i;
    for (i = 0; i < entity.length; ++i) {
      // increase fitness for each character that matches
      if (entity[i] == this.userData["solution"][i]) fitness += 1;

      // award fractions of a point as we get warmer
      fitness +=
        (127 -
          Math.abs(
            entity.charCodeAt(i) - this.userData["solution"].charCodeAt(i)
          )) /
        50;
    }
    return fitness;
    // return new Promise(function(resolve, reject) {
    //   setTimeout(resolve.bind(null, fitness), 100);
    // });
  }

  shouldContinue({ population: pop }: Genetic.GeneticState<Entity>) {
    // stop running once we've reached the solution
    return pop[0].entity != this.userData["solution"];
  }

  notification({
    population,
    isFinished,
    generation,
    stats,
  }: Genetic.Notification<Entity>) {
    if (isFinished) {
      console.log(
        JSON.stringify({ solution: population[0], generation, stats }, null, 2)
      );
    }
  }
}

const config: Partial<Genetic.Configuration> = {
  iterations: 4000,
  size: 250,
  crossover: 0.3,
  mutation: 0.3,
  skip: 20,
};

const userData: UserData = {
  solution:
    "Insanity is doing the same thing over and over again and expecting different results",
};
const genetic = new PhraseGenetic(config, userData);
genetic.evolve();
