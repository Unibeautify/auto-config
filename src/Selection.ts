// tslint:disable:no-function-expression object-literal-sort-keys
import { Genetic } from "@glavin001/genetic-js";
import {
  //   SingleSelection,
  //   PairWiseSelection,
  Population,
  Select1 as BaseSelect1,
  Select2 as BaseSelect2,
} from "@glavin001/genetic-js/dist/src/Selection";
import * as _ from "lodash";
import * as stringify from "json-stable-stringify";
import { Entity as UnibeautifyEntity } from "./UnibeautifySolver";

export const Select1 = {
  ...BaseSelect1,
};

export const Select2 = {
  ...BaseSelect2,
  RandomLinearRankAndRandom: function<Entity>(
    this: Genetic<Entity, any>,
    pop: Population<Entity>
  ): [Entity, Entity] {
    return [
      Select1.RandomLinearRank.call(this, pop),
      Select1.Random.call(this, pop),
    ];
  },
  RandomLinearRankAndTournament3: function<Entity>(
    this: Genetic<Entity, any>,
    pop: Population<Entity>
  ): [Entity, Entity] {
    return [
      Select1.RandomLinearRank.call(this, pop),
      Select1.Tournament3.call(this, pop),
    ];
  },
  RandomFittestAndTournament3: function<Entity>(
    this: Genetic<Entity, any>,
    pop: Population<Entity>
  ): [Entity, Entity] {
    const bestFitness = pop[0].fitness;
    const bestEntities = pop.filter(popItem => popItem.fitness === bestFitness);
    return [
      Select1.Random.call(this, bestEntities),
      Select1.Tournament3.call(this, pop),
    ];
  },
  DistinctPair: function<Entity>(
    this: Genetic<Entity, any>,
    pop: Population<Entity>
  ): [Entity, Entity] {
    const uniqPop = _.uniqBy(pop, ({ entity }) => stringify(entity));
    const items = _.sampleSize(uniqPop, 2);
    return [items[0].entity, items[1].entity];
    // const shuffled = _.shuffle(uniqPop);
    // return [shuffled[0], shuffled[1]];
  },
  MostDifferentPair: function<Entity>(
    this: Genetic<Entity, any>,
    pop: Population<Entity>
  ): [Entity, Entity] {
    const uniqPop = _.uniqBy(pop, ({ entity }) => stringify(entity));
    const firstEntity: Entity = Select1.Tournament3.call(this, uniqPop);
    const firstOptionsKeys = enabledOptionKeys(firstEntity as any);
    const sortedPop = _.sortBy(uniqPop, popItem => {
      return _.difference(enabledOptionKeys(popItem as any), firstOptionsKeys)
        .length;
    });
    const secondEntity = sortedPop[sortedPop.length - 1].entity;
    return [firstEntity, secondEntity];
  },
  DifferentPair: function<Entity>(
    this: Genetic<Entity, any>,
    pop: Population<Entity>
  ): [Entity, Entity] {
    const uniqPop = _.uniqBy(pop, ({ entity }) => stringify(entity));
    const firstEntity: Entity = Select1.Tournament3.call(this, uniqPop);
    const firstOptionsKeys = enabledOptionKeys(firstEntity as any);
    let diffPop = _.filter(uniqPop, popItem => {
      return (
        _.difference(firstOptionsKeys, enabledOptionKeys(popItem as any))
          .length !== 0
      );
    });
    if (diffPop.length === 0) {
      diffPop = _.filter(uniqPop, popItem => {
        return _.difference(enabledOptionKeys(popItem as any)).length !== 0;
      });
    }
    if (diffPop.length === 0) {
      diffPop = uniqPop;
    }
    const secondEntity = Select1.Tournament3.call(this, diffPop);
    return [firstEntity, secondEntity];
  },
  DifferentPair2: function<Entity>(
    this: Genetic<Entity, any>,
    pop: Population<Entity>
  ): [Entity, Entity] {
    const uniqPop = _.uniqBy(pop, ({ entity }) => stringify(entity));
    const firstEntity: Entity = Select1.Tournament3.call(this, uniqPop);
    const firstOptionsKeys = enabledOptionKeys(firstEntity as any);
    let diffPop = _.filter(uniqPop, popItem => {
      return (
        _.difference(enabledOptionKeys(popItem as any), firstOptionsKeys)
          .length !== 0
      );
    });
    if (diffPop.length === 0) {
      diffPop = _.filter(uniqPop, popItem => {
        return (
          _.difference(firstOptionsKeys, enabledOptionKeys(popItem as any))
            .length !== 0
        );
      });
    }
    if (diffPop.length === 0) {
      diffPop = uniqPop;
    }
    const secondEntity = Select1.Tournament3.call(this, diffPop);
    return [firstEntity, secondEntity];
  },
};

function enabledOptionKeys(entity: UnibeautifyEntity): string[] {
  return Object.keys(enabledOptions(entity));
}

function enabledOptions(entity: UnibeautifyEntity) {
  const originalOptions = entity.options;
  return _.omitBy(
    { ...originalOptions, beautifiers: undefined },
    value => value == undefined
  );
}
