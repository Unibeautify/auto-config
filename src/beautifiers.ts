import { Beautifier } from "unibeautify";
// import prettyDiff from "@unibeautify/beautifier-prettydiff";
import prettier from "@unibeautify/beautifier-prettier";
import jsBeautify from "@unibeautify/beautifier-js-beautify";
// import eslint from "@unibeautify/beautifier-eslint";
// import phpCsFixer from "@unibeautify/beautifier-php-cs-fixer";
// import clangFormat from "@unibeautify/beautifier-clang-format";
// import phpCodesniffer from "@unibeautify/beautifier-php-codesniffer";
// import black from "@unibeautify/beautifier-black";
// import gofmt from "@unibeautify/beautifier-gofmt";

const beautifiers: Beautifier[] = <any[]>[
  prettier,
  jsBeautify,
//   prettyDiff,
//   eslint,
//   phpCsFixer,
//   clangFormat,
//   phpCodesniffer,
//   black,
//   gofmt,
];

export default beautifiers;
