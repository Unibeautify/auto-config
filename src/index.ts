import * as Genetic from "genetic-js";

type Entity = string;
const genetic = Genetic.create<Entity>();
genetic.optimize = Genetic.Optimize.Maximize;
genetic.select1 = Genetic.Select1.Tournament2;
genetic.select2 = Genetic.Select2.Tournament2;

genetic.seed = function() {
  function randomString(len: number) {
    let text = "";
    const charset = "abcdefghijklmnopqrstuvwxyz0123456789";
    for (var i = 0; i < len; i++)
      text += charset.charAt(Math.floor(Math.random() * charset.length));

    return text;
  }

  // create random strings that are equal in length to solution
  return randomString(this.userData["solution"].length);
};

genetic.mutate = function(entity: Entity) {
  function replaceAt(str: string, index: number, character: string) {
    return (
      str.substr(0, index) + character + str.substr(index + character.length)
    );
  }

  // chromosomal drift
  var i = Math.floor(Math.random() * entity.length);
  return replaceAt(
    entity,
    i,
    String.fromCharCode(
      entity.charCodeAt(i) + (Math.floor(Math.random() * 2) ? 1 : -1)
    )
  );
};

genetic.crossover = function(mother: Entity, father: Entity) {
  // two-point crossover
  var len = mother.length;
  var ca = Math.floor(Math.random() * len);
  var cb = Math.floor(Math.random() * len);
  if (ca > cb) {
    var tmp = cb;
    cb = ca;
    ca = tmp;
  }

  var son =
    father.substr(0, ca) + mother.substr(ca, cb - ca) + father.substr(cb);
  var daughter =
    mother.substr(0, ca) + father.substr(ca, cb - ca) + mother.substr(cb);

  return [son, daughter];
};

genetic.fitness = function(entity: Entity) {
  var fitness = 0;

  var i;
  for (i = 0; i < entity.length; ++i) {
    // increase fitness for each character that matches
    if (entity[i] == this.userData["solution"][i]) fitness += 1;

    // award fractions of a point as we get warmer
    fitness +=
      (127 -
        Math.abs(
          entity.charCodeAt(i) - this.userData["solution"].charCodeAt(i)
        )) /
      50;
  }
  return fitness;
};

genetic.generation = function(
  pop: GeneticJs.Population<Entity>,
  generation: number,
  stats: GeneticJs.Stats
) {
  // stop running once we've reached the solution
  return pop[0].entity != this.userData["solution"];
};

genetic.notification = function(
  pop: GeneticJs.Population<Entity>,
  generation: number,
  stats: GeneticJs.Stats,
  isFinished: boolean
) {
  if (isFinished) {
    console.log(
      JSON.stringify({ solution: pop[0], generation, stats }, null, 2)
    );
  }
};

const config = {
  iterations: 4000,
  size: 250,
  crossover: 0.3,
  mutation: 0.3,
  skip: 20,
};

const userData = {
  solution:
    "Insanity is doing the same thing over and over again and expecting different results",
};

genetic.evolve(config, userData);
