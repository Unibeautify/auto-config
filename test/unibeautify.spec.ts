import * as Genetic from "@glavin001/genetic-js";
import prettier from "@unibeautify/beautifier-prettier";
import jsBeautify from "@unibeautify/beautifier-js-beautify";

import { raw } from "./raw-serializer";
import { UnibeautifyGenetic, UserData, Entity } from "../src/index";

describe("evolve", () => {
  testGenetic("pragma_insert", {
    beautifiers: [prettier, jsBeautify],
    language: "JavaScript",
    originalText: `console.log("hello world");`,
    desiredText: `/** @format */\n\nconsole.log('hello world')\n`,
  });

  testGenetic("arrow_parens", {
    beautifiers: [prettier, jsBeautify],
    language: "JavaScript",
    originalText: `a => {};`,
    desiredText: `(a) => {};`,
  });

  testGenetic("comma_first", {
    beautifiers: [prettier, jsBeautify],
    language: "JavaScript",
    originalText: `const a = "a", b = "b", c = "c";`,
    desiredText: `const a = "a"\n  , b = "b"\n  , c = "c";`,
  });

  testGenetic("test1", {
    beautifiers: [prettier, jsBeautify],
    language: "JavaScript",
    originalText: `if(true){console.log({ hello: "world" });}`,
    desiredText: `if(true){console.log({ hello: "world" });}`,
  });

  testGenetic("quotes + object_curly_spacing", {
    beautifiers: [prettier, jsBeautify],
    language: "JavaScript",
    originalText: `if(true){console.log({ hello: "world" });}`,
    desiredText: `if (true) {\n  console.log({hello: 'world'})\n}\n`,
  });

  testGenetic("indent_size + quotes + object_curly_spacing", {
    beautifiers: [prettier, jsBeautify],
    language: "JavaScript",
    originalText: `if(true){console.log({ hello: "world" });}`,
    desiredText: `if (true) {\n  console.log({hello: 'world'})\n}\n`,
  });

  testGenetic("quotes", {
    beautifiers: [prettier, jsBeautify],
    language: "JavaScript",
    originalText: `if(true){console.log("hello world");}`,
    desiredText: `if (true) {\n  console.log('hello world')\n}\n`,
  });

  testGenetic("object_curly_spacing", {
    beautifiers: [prettier, jsBeautify],
    language: "JavaScript",
    originalText: `var obj = { foo: "bar" };`,
    desiredText: `var obj = {foo: "bar"};\n`,
  });

  // testGenetic("end_with_comma", {
  //   beautifiers: [jsBeautify],
  //   language: "JavaScript",
  //   originalText: `var bar = {bar: "baz", qux: "quux"};`,
  //   desiredText: `var bar = {\n    bar: "baz",\n    qux: "quux",\n};`,
  // });

  // testGenetic("end_with_comma2", {
  //   beautifiers: [jsBeautify],
  //   language: "JavaScript",
  //   originalText: `var bar = {bar: "baz", qux: "quux"};\nvar foo = {bar: "baz", qux: "quux"};`,
  //   desiredText: `var bar = {\n  bar: "baz",\n  qux: "quux",\n};\nvar foo = {\n  bar: "baz",\n  qux: "quux",\n};`,
  // });

  testGenetic("break_chained_methods=true", {
    beautifiers: [jsBeautify],
    language: "JavaScript",
    originalText: `foo.bar().baz();`,
    desiredText: `foo.bar()\n  .baz();`,
  });

  // testGenetic("break_chained_methods=false", {
  //   beautifiers: [jsBeautify],
  //   language: "JavaScript",
  //   originalText: `foo.bar().baz();`,
  //   desiredText: `foo.bar().baz();`,
  // });

  testGenetic("break_chained_methods1", {
    beautifiers: [prettier, jsBeautify],
    language: "JavaScript",
    originalText: `this.$("#fileName").val().addClass("disabled")\n  .prop("disabled", true)`,
    desiredText: `this.$("#fileName")\n  .val()\n  .addClass("disabled")\n  .prop("disabled", true)\n`,
  });

  testGenetic("break_chained_methods2", {
    beautifiers: [prettier, jsBeautify],
    language: "JavaScript",
    originalText: `this.$("#fileName").val().addClass("disabled")\n  .prop("disabled", true)`,
    desiredText: `this.$("#fileName").val().addClass("disabled")\n  .prop("disabled", true)`,
  });
});

function testGenetic(title: string, userData: UserData) {
  const configuration: Partial<Genetic.Configuration> = {
    iterations: 100,
    size: 50,
    crossover: 0.8,
    mutation: 0.8,
    skip: 100,
  };
  test(title, done => {
    expect.hasAssertions();
    const genetic = new UnibeautifyGenetic({
      configuration,
      userData,
    });
    genetic.notification = function({
      population,
      isFinished,
    }: Genetic.Notification<Entity>) {
      if (isFinished) {
        this.beautify(population[0].entity).then((beautifiedText: string) => {
          expect(raw(beautifiedText)).toMatchSnapshot("beautifiedText");
          expect(population[0].entity).toMatchSnapshot("entity");
          expect(population[0].fitness).toMatchSnapshot("fitness");
          // expect(population[0]).toMatchSnapshot('');
          // expect(population[0].entity).toMatchSnapshot("entity");
          done();
        });
      }
    };
    return genetic.evolve().catch(error => {
      console.error(error);
      done();
    });
  });
}
