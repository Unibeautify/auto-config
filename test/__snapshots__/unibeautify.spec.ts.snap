// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`evolve arrow_parens: beautifiedText 1`] = `(a) => {};`;

exports[`evolve arrow_parens: entity 1`] = `
Object {
  "options": Object {
    "arrow_parens": "always",
    "beautifiers": Array [
      "Prettier",
      "JS-Beautify",
    ],
  },
}
`;

exports[`evolve arrow_parens: fitness 1`] = `1000`;

exports[`evolve break_chained_methods=false: beautifiedText 1`] = `foo.bar().baz();`;

exports[`evolve break_chained_methods=false: entity 1`] = `
Object {
  "options": Object {
    "beautifiers": Array [
      "JS-Beautify",
    ],
    "wrap_line_length": 160,
  },
}
`;

exports[`evolve break_chained_methods=false: fitness 1`] = `0`;

exports[`evolve break_chained_methods=true: beautifiedText 1`] = `
foo.bar()
  .baz();
`;

exports[`evolve break_chained_methods=true: entity 1`] = `
Object {
  "options": Object {
    "beautifiers": Array [
      "JS-Beautify",
    ],
    "break_chained_methods": true,
    "indent_size": 2,
  },
}
`;

exports[`evolve break_chained_methods=true: fitness 1`] = `1`;

exports[`evolve break_chained_methods1: beautifiedText 1`] = `
this.$("#fileName")
  .val()
  .addClass("disabled")
  .prop("disabled", true)

`;

exports[`evolve break_chained_methods1: entity 1`] = `
Object {
  "options": Object {
    "beautifiers": Array [
      "Prettier",
    ],
    "end_with_semicolon": false,
  },
}
`;

exports[`evolve break_chained_methods1: fitness 1`] = `0`;

exports[`evolve break_chained_methods2: beautifiedText 1`] = `
this.$("#fileName").val().addClass("disabled")
  .prop("disabled", true)
`;

exports[`evolve break_chained_methods2: entity 1`] = `
Object {
  "options": Object {
    "beautifiers": Array [
      "Prettier",
    ],
    "pragma_require": true,
  },
}
`;

exports[`evolve break_chained_methods2: fitness 1`] = `0`;

exports[`evolve comma_first: beautifiedText 1`] = `
const a = "a"
  , b = "b"
  , c = "c";
`;

exports[`evolve comma_first: entity 1`] = `
Object {
  "options": Object {
    "beautifiers": Array [
      "JS-Beautify",
    ],
    "comma_first": true,
    "indent_size": 2,
  },
}
`;

exports[`evolve comma_first: fitness 1`] = `1`;

exports[`evolve end_with_comma: beautifiedText 1`] = `
var bar = {
    bar: "baz",
    qux: "quux"
};

`;

exports[`evolve end_with_comma: entity 1`] = `
Object {
  "options": Object {
    "beautifiers": Array [
      "JS-Beautify",
    ],
    "end_with_newline": true,
  },
}
`;

exports[`evolve end_with_comma: fitness 1`] = `10000`;

exports[`evolve end_with_comma2: beautifiedText 1`] = `
var bar = {
  bar: "baz",
  qux: "quux",
};
var foo = {
  bar: "baz",
  qux: "quux",
};
`;

exports[`evolve end_with_comma2: entity 1`] = `
Object {
  "options": Object {
    "beautifiers": Array [
      "JS-Beautify",
    ],
    "end_with_comma": true,
    "indent_size": 2,
  },
}
`;

exports[`evolve end_with_comma2: fitness 1`] = `20001`;

exports[`evolve indent_size + quotes + object_curly_spacing: beautifiedText 1`] = `
if (true) {
  console.log({hello: 'world'})
}

`;

exports[`evolve indent_size + quotes + object_curly_spacing: entity 1`] = `
Object {
  "options": Object {
    "beautifiers": Array [
      "Prettier",
    ],
    "end_with_semicolon": false,
    "object_curly_spacing": false,
    "quotes": "single",
  },
}
`;

exports[`evolve indent_size + quotes + object_curly_spacing: fitness 1`] = `2`;

exports[`evolve object_curly_spacing: beautifiedText 1`] = `
var obj = {foo: "bar"};

`;

exports[`evolve object_curly_spacing: entity 1`] = `
Object {
  "options": Object {
    "beautifiers": Array [
      "Prettier",
    ],
    "object_curly_spacing": false,
  },
}
`;

exports[`evolve object_curly_spacing: fitness 1`] = `0`;

exports[`evolve pragma_insert: beautifiedText 1`] = `
/** @format */

console.log('hello world')

`;

exports[`evolve pragma_insert: entity 1`] = `
Object {
  "options": Object {
    "beautifiers": Array [
      "Prettier",
    ],
    "end_with_semicolon": false,
    "pragma_insert": true,
    "quotes": "single",
  },
}
`;

exports[`evolve pragma_insert: fitness 1`] = `2`;

exports[`evolve quotes + object_curly_spacing: beautifiedText 1`] = `
if (true) {
  console.log({hello: 'world'})
}

`;

exports[`evolve quotes + object_curly_spacing: entity 1`] = `
Object {
  "options": Object {
    "beautifiers": Array [
      "Prettier",
    ],
    "end_with_semicolon": false,
    "object_curly_spacing": false,
    "quotes": "single",
  },
}
`;

exports[`evolve quotes + object_curly_spacing: fitness 1`] = `2`;

exports[`evolve quotes: beautifiedText 1`] = `
if (true) {
  console.log('hello world')
}

`;

exports[`evolve quotes: entity 1`] = `
Object {
  "options": Object {
    "beautifiers": Array [
      "Prettier",
    ],
    "end_with_semicolon": false,
    "quotes": "single",
  },
}
`;

exports[`evolve quotes: fitness 1`] = `1`;

exports[`evolve test1: beautifiedText 1`] = `if(true){console.log({ hello: "world" });}`;

exports[`evolve test1: entity 1`] = `
Object {
  "options": Object {
    "beautifiers": Array [
      "Prettier",
    ],
    "pragma_require": true,
  },
}
`;

exports[`evolve test1: fitness 1`] = `0`;
